CRM/Civicase/Service/CaseTypeCategoryEventHandler.php

```diff
@@ -0,0 +1,236 @@
+<?php
+
+/**
+ * Class CRM_Civicase_Service_CaseTypeCategoryEventHandler.
+ */
+class CRM_Civicase_Service_CaseTypeCategoryEventHandler {
+
+  /**
+   * Menu handler.
+   *
+   * @var \CRM_Civicase_Service_CaseCategoryMenu
+   */
+  protected $menu;
+
+  /**
+   * Custom data handler.
+   *
+   * @var \CRM_Civicase_Service_CaseCategoryCustomDataType
+   */
+  protected $customData;
+
+  /**
+   * Custom field handler.
+   *
+   * @var \CRM_Civicase_Service_CaseCategoryCustomFieldExtends
+   */
+  protected $customFieldExtends;
+
+  /**
+   * Should we check referer before running action?
+   *
+   * @var bool
+   */
+  protected $needCheckReferer = FALSE;
+
+  /**
+   * Should we check option group before running action?
+   *
+   * Depends on $optionGroupId.
+   *
+   * @var bool
+   */
+  protected $needCheckOptionGroup = TRUE;
+
+  /**
+   * Case type category option group id. Used to check option group.
+   *
+   * @var int
+   */
+  protected $optionGroupId;
+
+  /**
+   * Object name.
+   *
+   * Is used to check object if it's type is unpredictable or unknown.
+   *
+   * @var string
+   */
+  protected $objectName = 'OptionValue';
+
+  /**
+   * CRM_Civicase_Service_CaseTypeCategoryEventHandler constructor.
+   *
+   * @param object $menu
+   *   Menu handler.
+   * @param object $customData
+   *   Custom data handler.
+   * @param object $customFieldExtends
+   *   Custom field handler.
+   */
+  public function __construct($menu, $customData, $customFieldExtends) {
+    $this->menu = $menu;
+    $this->customData = $customData;
+    $this->customFieldExtends = $customFieldExtends;
+  }
+
+  /**
+   * Perform actions on case type category create.
+   *
+   * @param string $name
+   *   Case type category name.
+   */
+  public function onCreate($name) {
+    if (!$name || !$this->shouldRun()) {
+      return;
+    }
+
+    $this->menu->createItems($name);
+    $this->customFieldExtends->create($name, "Case ({$name})");
+    $this->customData->create($name);
+  }
+
+  /**
+   * Perform actions on case type category create.
+   *
+   * @param array $category
+   *   Case type category data, should contain:
+   *   id - category id,
+   *   is_active - is_active category property (enabled/disabled),
+   *   icon - category icon.
+   */
+  public function onUpdate(array $category) {
```

Comment:
It is actually better to pass the parameters directly rather than as an array since we don't have that much to update for now. We will avoid all the if statements in the function and function will be cleaner.

```php
  public function OnUpdate($caseCategoryId, $caseCategoryIcon, $caseCategoryStatus){
    $updateParams = [
      'is_active' => !empty($caseCategoryStatus) ? 1 : 0,
      'icon' => 'crm-i ' . $caseCategoryIcon,
    ];
    $this->caseCategoryMenu->updateItems($caseCategoryId, $updateParams);
  }
```

Reply:
We would not avoid `if` statements, because post process hook and api listener would pass different number of parameters here. And `is_active` parameter may be not specified on some api calls.
Also this approach would save time for refactoring in the future: we would not need to search for all usages of this function and update them just because one more parameter was added.

Reply:
The function that the onUpdate class accepts should be constant. The Event Handler class is a service and it is the responsibility of those classes to format their requests to what this class expects. If we don't do that, then this class becomes a garbage that can be dumped anything and it will have to format the parameters itself.

Like I said we don't have more parameters to update now and we might not likely have more. It is actually not good practice to pass arrays to a function if it can be avoided. Passing the parameters separately allows clarity, typhinting ....